test/puppetlabs/http/client/async_unbuffered_test.clj

```diff
@@ -327,3 +328,167 @@
+                  (is (<= 1 (get metric-data "mean")))
+                  (is (<= 1 (get metric-data "aggregate")))
+
+                  (is (= 1 (get unbuffered-stream-data "count")))
+                  (is (= unbuffered-stream-id (get unbuffered-stream-data "metric-id")))
+                  (is (<= 1000 (get unbuffered-stream-data "mean")))
+                  (is (<= 1000 (get unbuffered-stream-data "aggregate")))
+
+                  (is (> (get unbuffered-stream-data "mean") (get metric-data "mean"))))))))))
+     (testing "metrics work for failed request"
+       (try
+         (testwebserver/with-test-webserver-and-config
+          (blocking-handler data) port {:shutdown-timeout-seconds 1}
+          (let [metric-registry (MetricRegistry.)]
+            (with-open [client (-> (ClientOptions.)
+                                   (.setSocketTimeoutMilliseconds 200)
+                                   (.setConnectTimeoutMilliseconds 100)
+                                   (Async/createClient metric-registry))]
+              (let [request-options (doto (RequestOptions. (str "http://localhost:" port "/hello"))
+                                      (.setAs ResponseBodyType/UNBUFFERED_STREAM))
+                    response (-> client (.get request-options) .deref)
+                    error (.getError response)
+                    body (.getBody response)]
+                (is (nil? error))
+                (is (thrown? SocketTimeoutException (slurp body)))
+                (let [client-metrics (.getClientMetrics client)
+                      client-metrics-data (.getClientMetricsData client)
+                      metric-id (str "puppetlabs.http-client.http://localhost:" port "/hello.GET")
+                      unbuffered-stream-id (str metric-id ".unbuffered_stream")]
+                  (is (= (set (list metric-id unbuffered-stream-id))
+                         (set (keys client-metrics))
+                         (set (keys client-metrics-data))))
+                  (is (every? #(instance? Timer %) (vals client-metrics)))
+                  (let [metric-data (get client-metrics-data metric-id)
+                        unbuffered-stream-data (get client-metrics-data unbuffered-stream-id)]
+                    (is (= 1 (get metric-data "count")))
+                    (is (= metric-id (get metric-data "metric-id")))
+                    (is (<= 1 (get metric-data "mean")))
+                    (is (<= 1 (get metric-data "aggregate")))
+
+                    (is (= 1 (get unbuffered-stream-data "count")))
+                    (is (= unbuffered-stream-id (get unbuffered-stream-data "metric-id")))
+                    (is (<= 1 (get unbuffered-stream-data "mean")))
+                    (is (<= 1 (get unbuffered-stream-data "aggregate")))
+
+                    (is (> (get unbuffered-stream-data "mean") (get metric-data "mean")))))))))
+         (catch TimeoutException e
+           ;; Expected whenever a server-side failure is generated
```

Comment:
What causes the server-side failure in this case?

Could we add a couple of pathological assertions `(is (true? false))` and `(is (true? true))` in the spots in the test code path that we do/don't expect it to reach due to the exception?